src/map_matrix.c

```diff
@@ -0,0 +1,231 @@
+    u16 matrix_id = MapHeader_GetMatrixId(map_no);
+
+    MapMatrix_MapMatrixData_Load(&map_matrix->data, matrix_id, map_no);
+
+    map_matrix->matrix_id = matrix_id;
+    map_matrix->height = map_matrix->data.height;
+    map_matrix->width = map_matrix->data.width;
+}
+
+void MapMatrix_Free(MAPMATRIX* map_matrix) {
+    FreeToHeap(map_matrix);
+}
+
+u16 MapMatrix_GetMapData(s32 map_no, MAPMATRIX* map_matrix) {
+    GF_ASSERT(map_no < map_matrix->width * map_matrix->height);
+    return map_matrix->data.maps.data[map_no];
+}
+
+u8 MapMatrix_GetWidth(MAPMATRIX* map_matrix) {
+    GF_ASSERT(map_matrix != NULL);
+    return map_matrix->width;
+}
+
+u8 MapMatrix_GetHeight(MAPMATRIX* map_matrix) {
+    GF_ASSERT(map_matrix != NULL);
+    return map_matrix->height;
+}
+
+u16 MapMatrix_GetMapHeader(MAPMATRIX* map_matrix, s32 x, s32 y) {
+    s32 width = map_matrix->width;
+    s32 height = map_matrix->height;
+
+    GF_ASSERT(x >= 0 && x < width);
+    GF_ASSERT(y >= 0 && y < height);
+
+    return map_matrix->data.headers[y * width + x];
+}
+
+u8 MapMatrix_GetMatrixId(MAPMATRIX* map_matrix) {
+    return map_matrix->matrix_id;
+}
+
+u8 MapMatrix_GetMapAltitude(MAPMATRIX* map_matrix, u8 matrix_id, u16 x, u16 y, int matrix_width) {
+#pragma unused(matrix_id)
+    GF_ASSERT(x < matrix_width);
+    GF_ASSERT(y * matrix_width + x < MAP_MATRIX_MAX_SIZE);
+
+    return map_matrix->data.altitudes[y * matrix_width + x];
+}
+
+MAPDATA* MapMatrix_MapData_New(u32 heap_id) {
+    MAPDATA* map_data = AllocFromHeap(heap_id, sizeof(MAPDATA));
+
+    void* buffer = AllocAtEndAndReadWholeNarcMemberByIdPair(NARC_fielddata_mapmatrix_map_matrix, 0, heap_id);
+    u8* cursor = (u8*)buffer;
+    cursor += 4;
+    u8 name_length = *cursor;
+    cursor++;
+    cursor += name_length;
+
+    MI_CpuCopy8(cursor, map_data, sizeof(MAPDATA));
+    FreeToHeap(buffer);
+
+    return map_data;
+}
+
+void MapMatrix_MapData_Free(MAPDATA* map_data) {
+    GF_ASSERT(map_data != NULL);
+    FreeToHeap(map_data);
+}
+
+u16 GetMapData(u32 map_no, MAPMATRIX* map_matrix) {
+    GF_ASSERT(map_matrix != NULL);
+    return MapMatrix_GetMapData(map_no, map_matrix);
+}
+
+void RemoveMahoganyTownAntennaTree(MAPMATRIX* map_matrix) {
+    u16* maps = map_matrix->data.maps.data;
+    u8 width = map_matrix->width;
+
+    if (map_matrix->matrix_id != NARC_map_matrix_map_matrix_00000000_bin) {
+        return;
+    }
+
+    maps[width * 5 + 16] = 86;
+}
+
+BOOL ShouldUseAlternateLakeOfRage(SAVEDATA* savedata, u32 map_no) {
+    RTCDate date;
+    struct ScriptState *state = SavArray_Flags_get(savedata);
+
+    GF_RTC_CopyDate(&date);
+
+    if (CheckFlagInArray(state, 202) == FALSE) {
+        // The player hasn't battled the Red Gyarados yet.
+        sub_02066C4C(state, 1);
+        return FALSE;
+    }
+
+    // FIXME(tgsm): this smells like a fakematch, but honestly I'm not sure.
+    if ((map_no == MAP_T29 || map_no == MAP_R43) && date.week == RTC_WEEK_WEDNESDAY) {
```

Comment:
I can't remember atm but have you tried making this conditional a static inline?

Reply:
Yup, that was it.